In [ ]:
### Note
* Instructions have been included for each segment.
You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np

In [2]:
# File to Load (Remember to Change These)
file = "Resources/purchase_data.csv"

In [3]:
# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [4]:
total_players=len(purchase_data.SN.unique())
#define a data fram for the total players
total_players_df= pd.DataFrame({"Total Players":[total_players]})
#display the Dataframe 
total_players_df

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [5]:
#defining all of the variables in a cleaner format
unique_items = len(purchase_data["Item ID"].unique())
Average_Price =round(purchase_data["Price"].mean(),2)
total_purchases = purchase_data["Item ID"].count()
Total_price = purchase_data["Price"].sum()
Max_Price =round(purchase_data["Price"].max(),2)
Min_Price =round(purchase_data["Price"].min(),2)

#summarizing the data into a data frame
summary_of_data = pd.DataFrame({
    "Number of unique items":[unique_items],
    "Average price":[Average_Price],
    "Total purchase":[total_purchases],
    "Maximum price":[Max_Price],
    "Minimum price":[Min_Price],
    "Revenue":[Total_price] })

#formatting our data summary
summary_of_data["Average price"] = summary_of_data["Average price"].map("${:.2f}".format)
summary_of_data["Revenue"] = summary_of_data["Revenue"].map("${:.2f}".format)
summary_of_data["Maximum price"] = summary_of_data["Maximum price"].map("${:.2f}".format)
summary_of_data["Minimum price"] = summary_of_data["Minimum price"].map("${:.2f}".format)

summary_of_data

Number of unique items Average price  Total purchase Maximum price  \
0                     183         $3.05             780         $4.99   

  Minimum price   Revenue  
0         $1.00  $2379.77

In [ ]:
#another way of showing the summary of the purchasing analysis
purchase_analysis= pd.DataFrame(round(purchase_data.describe(),2))
purchase_analysis

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [6]:
gender = pd.DataFrame(purchase_data[["SN","Gender"]].drop_duplicates("SN"))
gender = pd.DataFrame(gender.groupby("Gender")["Gender"].count())
gender = gender.rename(columns={"Gender":"Players by Gender"})


players = pd.DataFrame(round((gender["Players by Gender"] / total_players)*100,2).map("{0:.2f}%".format))
players= players.rename(columns={"Players by Gender":"Percentage of players by Gender"})

demographic= [gender,players]
gender_demographic= pd.concat(demographic,axis=1,join='inner')
gender_demographic = gender_demographic.sort_values("Percentage of players by Gender", ascending=False) 
gender_demographic

Players by Gender Percentage of players by Gender
Gender                                                                  
Male                                 484                          84.03%
Female                                81                          14.06%
Other / Non-Disclosed                 11                           1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:

total_purchase_bygender = purchase_data.groupby(["Gender"]).sum()["Price"].rename("Total Purchase Value")
Average_bygender = purchase_data.groupby(["Gender"]).mean()["Price"].rename("Average Purchase Price")
count_bygender = purchase_data.groupby(["Gender"]).count()["Price"].rename("Purchase Count")

# Calculatingt the Avg Total Purchase per Person
Average_purchase_perperson= total_purchase_bygender / gender_demographic["Players by Gender"]


purchasing_analysis = pd.DataFrame({
    "Purchase Count": count_bygender , 
    "Average Purchase Price": Average_bygender, 
    "Total Purchase Value": total_purchase_bygender, 
    "Avg Total Purchase per Person": Average_purchase_perperson})

#formating
purchasing_analysis["Average Purchase Price"] = purchasing_analysis["Average Purchase Price"].map("${:,.2f}".format)
purchasing_analysis["Total Purchase Value"] = purchasing_analysis["Total Purchase Value"].map("${:,.2f}".format)
purchasing_analysis ["Purchase Count"] = purchasing_analysis["Purchase Count"].map("{:,}".format)
purchasing_analysis["Avg Total Purchase per Person"] = purchasing_analysis["Avg Total Purchase per Person"].map("${:,.2f}".format)


purchasing_analysis = purchasing_analysis.loc[:, ["Purchase Count", "Average Purchase Price", "Total Purchase Value", "Avg Total Purchase per Person"]]
purchasing_analysis

Purchase Count Average Purchase Price  \
Gender                                                        
Female                           113                  $3.20   
Male                             652                  $3.02   
Other / Non-Disclosed             15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [8]:
age_list = purchase_data[["SN","Age","Price"]]
age_list.head()

SN  Age  Price
0        Lisim78   20   3.53
1    Lisovynya38   40   1.56
2     Ithergue48   24   4.88
3  Chamassasya86   24   3.27
4      Iskosia90   23   1.44

In [9]:
# Bins for ages and group names
Ages= [0, 9.90, 14.90, 19.90, 24.90, 29.90, 34.90, 39.90, 99999]
age_groups= ["under 10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "over 40"]

#Putting Bin values into discrete intervals
age_list["Age Group"] = pd.cut(age_list["Age"], Ages, labels = age_groups)
age_list.head()

C:\Users\dshakoory\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


SN  Age  Price Age Group
0        Lisim78   20   3.53     20-24
1    Lisovynya38   40   1.56   over 40
2     Ithergue48   24   4.88     20-24
3  Chamassasya86   24   3.27     20-24
4      Iskosia90   23   1.44     20-24

In [10]:
#count by our age groups
age_demo_table = age_list.groupby("Age Group").SN.nunique().reset_index(name='Total Count')

#calculating the percentage of players cosnidering the count by age group
percentage = round((age_demo_table["Total Count"] /total_players)*100,2)
age_demo_table["Percentage of Players"] = percentage
age_demo_table = age_demo_table.set_index("Age Group")
#changing the format of data to percentage 
age_demo_table["Percentage of Players"] = age_demo_table ["Percentage of Players"].map("${:.2f}%".format) 

#Displaying the data frame
age_demo_table 

Total Count Percentage of Players
Age Group                                   
under 10            17                $2.95%
10-14               22                $3.82%
15-19              107               $18.58%
20-24              258               $44.79%
25-29               77               $13.37%
30-34               52                $9.03%
35-39               31                $5.38%
over 40             12                $2.08%

## Purchasing Analysis (Age)

Bin the purchase_data data frame by age
Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below
Create a summary data frame to hold the results
Optional: give the displayed data cleaner formatting
Display the summary data frame

In [11]:
#calculating the average purchase price using the mean in the data frame
Avg_price_byage = pd.DataFrame(age_list.groupby("Age Group")["Price"].mean())
#renaming the calculated column
Avg_price_byage = Avg_price_byage.rename(columns={"Price":"Average Purchase Price"})
#changing the format of data to currency and cleaning the data
Avg_price_byage["Average Purchase Price"] = Avg_price_byage["Average Purchase Price"].map("${:.2f}".format)
#Displaying the data frame
Avg_price_byage 


Average Purchase Price
Age Group                       
under 10                   $3.35
10-14                      $2.96
15-19                      $3.04
20-24                      $3.05
25-29                      $2.90
30-34                      $2.93
35-39                      $3.60
over 40                    $2.94

In [12]:
#calculating the count of purchases using the .count in the data frame
purch_cnt_byage = pd.DataFrame(age_list.groupby("Age Group")["SN"].count())
#renaming the calculated column
purch_cnt_byage = purch_cnt_byage.rename(columns={"SN":"Total Purchases"})
#Displaying the data frame
purch_cnt_byage 

Total Purchases
Age Group                 
under 10                23
10-14                   28
15-19                  136
20-24                  365
25-29                  101
30-34                   73
35-39                   41
over 40                 13

In [13]:
#calculating the total of purchases using the .sum in the data frame
total_purch_byage = pd.DataFrame(age_list.groupby("Age Group")["Price"].sum())
#renaming the calculated column
total_purch_byage  = pd.DataFrame(total_purch_byage .rename(columns={"Price":"Total Purchase Value"}))
#changing the format of data and cleaning the data
total_purch_byage  = pd.DataFrame(total_purch_byage ["Total Purchase Value"].map("${:.2f}".format))
#Displaying the data frame
total_purch_byage 

Total Purchase Value
Age Group                     
under 10                $77.13
10-14                   $82.78
15-19                  $412.89
20-24                 $1114.06
25-29                  $293.00
30-34                  $214.00
35-39                  $147.67
over 40                 $38.24

In [14]:
#calculating the average total purchases per person 
avg_totalpurch_perperson=pd.DataFrame(age_list.groupby("Age Group")["Price"].sum() /age_list.groupby("Age Group").SN.nunique())
#renaming the calculated column
aavg_totalpurch_perperson=avg_totalpurch_perperson.rename(columns={0:"Average Total Purchase per person"}, inplace="true")

#changing the format of data and cleaning the data
avg_totalpurch_perperson=pd.DataFrame(avg_totalpurch_perperson ["Average Total Purchase per person"].map("${:.2f}".format))
#Displaying the data frame
avg_totalpurch_perperson


Average Total Purchase per person
Age Group                                  
under 10                              $4.54
10-14                                 $3.76
15-19                                 $3.86
20-24                                 $4.32
25-29                                 $3.81
30-34                                 $4.12
35-39                                 $4.76
over 40                               $3.19

In [15]:
#The best way is to add all of our code into one section and merge the columns on our final data frame to display
#we have included the process of data cleaning and munging during this process as well
Avg_price_byage = pd.DataFrame(age_list.groupby("Age Group")["Price"].mean())
Avg_price_byage = Avg_price_byage.rename(columns={"Price":"Average Purchase Price"})
Avg_price_byage["Average Purchase Price"] = Avg_price_byage["Average Purchase Price"].map("${:.2f}".format)

purch_cnt_byage = pd.DataFrame(age_list.groupby("Age Group")["SN"].count())
purch_cnt_byage = purch_cnt_byage.rename(columns={"SN":"Total Purchases"})

total_purch_byage = pd.DataFrame(age_list.groupby("Age Group")["Price"].sum())
total_purch_byage  = pd.DataFrame(total_purch_byage .rename(columns={"Price":"Total Purchase Value"}))
total_purch_byage  = pd.DataFrame(total_purch_byage ["Total Purchase Value"].map("${:.2f}".format))

purchasing_analysis_byage =pd.merge(pd.merge(purch_cnt_byage, Avg_price_byage,on="Age Group"),total_purch_byage,on="Age Group")

purchasing_analysis_byage['Average Total Purchase perperson']= pd.DataFrame(age_list.groupby("Age Group")["Price"].sum() /age_list.groupby("Age Group").SN.nunique())
purchasing_analysis_byage['Average Total Purchase perperson']=pd.DataFrame(purchasing_analysis_byage['Average Total Purchase perperson'].map("${:.2f}".format))
purchasing_analysis_byage

Total Purchases Average Purchase Price Total Purchase Value  \
Age Group                                                                
under 10                23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05             $1114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
over 40                 13                  $2.94               $38.24   

          Average Total Purchase perperson  
Age Group                                   
under 10                             $4.54  
10-14                                $3.76  
15-19                                $3.86  
20-24                                $4.32  
25-29                                $3.81  
30-34                                $4.12  
35-39                                $4.76  
over 40                              $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [16]:
#manipulating the data,Changing the order to Ascending,Changing the format in the data fram and renaming the column
top_spenders = pd.DataFrame(purchase_data.groupby("SN")["Price"].sum())
top_spenders = pd.DataFrame(top_spenders.sort_values(by="Price", ascending=False))
top_spenders = pd.DataFrame(top_spenders["Price"].map("${:.2f}".format))
top_spenders = pd.DataFrame(top_spenders.rename(columns = {"Price":"Total Purchase Value"}))

#manipulating the data,Changing the order to Ascending,Changing the format in the data fram and renaming the column
top_spender_ct = pd.DataFrame(purchase_data.groupby("SN")["Item Name"].count())
top_spender_ct = top_spender_ct.sort_values(by="Item Name", ascending=False)
top_spender_ct = pd.DataFrame(top_spender_ct.rename(columns = {"Item Name":"Purchase Count"}))

#building our data frame based on the manipulation before
top_spender_avgprice = pd.DataFrame(purchase_data.groupby("SN")["Price"].mean())
top_spender_avgprice = pd.DataFrame(top_spender_avgprice["Price"].map("${:.2f}".format))
top_spender_avgprice = pd.DataFrame(top_spender_avgprice.rename(columns = {"Price":"Average Purchase Price"}))

#merging the column with Concat with inner join on SN
Top_spender_list= [top_spender_ct,top_spenders,top_spender_avgprice]
Top_spender_list= pd.concat(Top_spender_list,axis=1,join='inner')
Top_spender_list.head()

Purchase Count Total Purchase Value Average Purchase Price
SN                                                                     
Lisosia93                 5               $18.96                  $3.79
Iral74                    4               $13.62                  $3.40
Idastidru52               4               $15.45                  $3.86
Asur53                    3                $7.44                  $2.48
Inguron55                 3               $11.11                  $3.70

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [17]:

#first step is manipulating the data
item_group= purchase_data.groupby(["Item ID", "Item Name"])
item_group_count = item_group["SN"].count()
item_group_total = item_group["Price"].sum()
item_price = (item_group_total / item_group_count)
item_value = (item_price * item_group_count)

#Formatting our data fram after data manipulation 
Popular_Items = pd.DataFrame({
    "Purchase Count": item_group_count,
    "Item Price": item_price ,
    "Total Purchase Value": item_value})

Popular_Items = Popular_Items.sort_values("Purchase Count", ascending=False) 
Popular_Items["Item Price"] = Popular_Items["Item Price"].map("${:.2f}".format)
Popular_Items["Total Purchase Value"] = Popular_Items["Total Purchase Value"].map("${:.2f}".format)
Popular_Items = Popular_Items[["Purchase Count", "Item Price", "Total Purchase Value"]]
Popular_Items.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [18]:
#manipulating the data by using calculation functions
item_group= purchase_data.groupby(["Item ID", "Item Name"])
item_group_countbygender = item_group["Gender"].count()
item_group_total = item_group["Price"].sum()
item_price = (item_group_total / item_group_countbygender)

#Defining New Data Frame
profitable_items = pd.DataFrame({"Purchase Count": item_group_countbygender,
                          "Item Price": item_price,
                          "Total Purchase Value": item_group_total})

#Sort the above table by total purchase value in descending order
profitable_items = profitable_items.sort_values("Total Purchase Value", ascending=False) 
profitable_items["Item Price"] = profitable_items["Item Price"].map("${:.2f}".format)
profitable_items["Total Purchase Value"] = profitable_items["Total Purchase Value"].map("${:.2f}".format)
profitable_items = profitable_items[["Purchase Count", "Item Price", "Total Purchase Value"]]
profitable_items.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

# Three observable trends based on the data

A) Purchasing Analysis by Gender
1. There are more Male players than Female or Other players according to the gender demographic table.However, The Purchasing Analysis by Gender indicates that average purchase price consumed by female players is higher than the male players.

B) Age Demographic 
2. Per the data redulted from the Age Demographic it is noted that the age group between 20 -24 holds the highest percent of players (followed by 15-19 then 25-29). 

3. Players under the age of 10 and over 40 years old are both less than 3%

C)Top Spenders
4. The moost profitable and the most popular game is “Oathbreaker, Last Hope of the Breaking Storm” and based on the top spender table it is notable that Lisosia93 is our highest spender for these games with total purchase amount of $18.96.

